samples/order/frontend.go

```diff
@@ -12,19 +12,193 @@
+func (*frontend) orderForm(w http.ResponseWriter, r *http.Request) error {
+	if r.Method != "GET" {
+		http.Error(w, "bad method for orderForm: want GET", http.StatusBadRequest)
+		return nil
+	}
+	return executeTemplate(orderFormTemplate, nil, w)
+}
+
+// createOrder handles a submitted order form.
+func (f *frontend) createOrder(w http.ResponseWriter, r *http.Request) error {
+	if r.Method != "POST" {
+		http.Error(w, "bad method for createOrder: want POST", http.StatusBadRequest)
+		return nil
+	}
+	email := r.FormValue("email")
+	if email == "" {
+		http.Error(w, "email missing", http.StatusBadRequest)
+		return nil
+	}
+
+	file, _, err := r.FormFile("file")
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	if _, err := f.doCreateOrder(r.Context(), email, file, time.Now()); err != nil {
+		return err
+	}
+	fmt.Fprintln(w, "Order received. Thank you.")
+	return nil
+}
+
+// doCreateOrder returns the order ID it generates, for testing.
+func (f *frontend) doCreateOrder(ctx context.Context, email string, file io.Reader, now time.Time) (id string, err error) {
+	// Assign an ID for the order here, rather than in the processor.
+	// That allows the processor to detect duplicate pub/sub messages.
+	id = f.newID()
+	req := &OrderRequest{
+		ID:         id,
+		InImage:    id + "-in",
+		Email:      email,
+		CreateTime: now,
+	}
+
+	// Copy the uploaded input file to the bucket.
+	w, err := f.bucket.NewWriter(ctx, req.InImage, nil)
+	if err != nil {
+		return "", err
+	}
+	defer func() {
+		err2 := w.Close()
+		if err == nil {
+			err = err2
+		}
+	}()
+
+	_, err = io.Copy(w, file)
+	if err != nil {
+		return "", err
+	}
+
+	defer func() {
+		// if we can't send the request, the image will never be processed.
+		// Try to delete it.
+		if err != nil {
+			if err := f.bucket.Delete(ctx, req.InImage); err != nil {
+				log.Printf("deleting orphan image %q: %v", req.InImage, err)
+			}
+		}
+	}()
+
+	// Publish the new order.
+	bytes, err := json.Marshal(req)
+	if err != nil {
+		return "", err
+	}
+	if err := f.requestTopic.Send(ctx, &pubsub.Message{Body: bytes}); err != nil {
+		return "", err
+	}
+	return id, nil
+}
+
+// listOrders lists all the orders in the database.
+func (f *frontend) listOrders(w http.ResponseWriter, r *http.Request) error {
+	if r.Method != "GET" {
+		http.Error(w, "bad method for listOrders: want GET", http.StatusBadRequest)
+		return nil
+	}
+	ctx := r.Context()
+	iter := f.coll.Query().Get(ctx)
+	var orders []*Order
+	for {
+		var ord Order
+		err := iter.Next(ctx, &ord)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		orders = append(orders, &ord)
+	}
+	return executeTemplate(listTemplate, orders, w)
+}
+
+// newID creates a new unique ID for an incoming order. It uses the current
+// second, formatted in a readable way. The resulting IDs sort nicely and are
+// easy to read, but of course are not suitable for production because there
```

Comment:
why not append a running integer to the second? will still sort nicely and likely won't take much more code than this comment :)

Reply:
You'll get dups on a fast restart or with clock resets. My concern is that we'd be giving the illusion that we're dealing with dups and making the code (slightly) more complex, but we haven't come any closer to a production-quality implementation. (In fact it's worse, because it will fail more rarely.)

I'd be open to a simple production-quality idea (like using a UUID) or something even dumber than what I have (an in-memory counter).

Reply:
Fair enough.